src/language-js/types/estree.d.ts

```diff
@@ -0,0 +1,55 @@
+import * as ESTree from "estree";
+import * as Babel from "@babel/types";
+import * as TSESTree from "@typescript-eslint/types/dist/ts-estree";
+import * as NGTree from "angular-estree-parser/lib/types";
+
+type AnyNode = ESTree.Node | Babel.Node | TSESTree.Node | NGTree.NGNode;
+
+export type TemplateLiteral =
+  | ESTree.TemplateLiteral
+  | Babel.TemplateLiteral
+  | TSESTree.TemplateLiteral;
+export type Comment = ESTree.Comment | Babel.Comment | TSESTree.Comment;
+export type CallExpression =
+  | ESTree.CallExpression
+  | Babel.CallExpression
+  | TSESTree.CallExpression;
+export type OptionalCallExpression =
+  | Babel.OptionalCallExpression
+  | TSESTree.OptionalCallExpression;
+export type MemberExpression =
+  | ESTree.MemberExpression
+  | Babel.MemberExpression
+  | TSESTree.MemberExpression;
+export type OptionalMemberExpression =
+  | Babel.OptionalMemberExpression
+  | TSESTree.OptionalMemberExpression;
+export type Expression =
+  | ESTree.Expression
+  | Babel.Expression
+  | TSESTree.Expression;
+export type BindExpression = Babel.BindExpression;
+export type Property = ESTree.Property | Babel.Property | TSESTree.Property;
+export type ClassPrivateProperty = Babel.ClassPrivateProperty;
+export type ObjectTypeProperty = Babel.ObjectTypeProperty;
+export type JSXElement = Babel.JSXElement | TSESTree.JSXElement;
+export type TaggedTemplateExpression =
+  | ESTree.TaggedTemplateExpression
+  | Babel.TaggedTemplateExpression
+  | TSESTree.TaggedTemplateExpression;
+export type Literal = ESTree.Literal | Babel.Literal | TSESTree.Literal;
+
+type PrettierEsComment = Comment & {
+  printed?: boolean;
+  trailing?: boolean;
+  leading?: boolean;
+};
+
+export type Node = AnyNode & {
+  extra?: any;
+  comments?: PrettierEsComment[];
+  trailingComments?: any;
+  // leadingComments?: PrettierEsComment[];
```

Comment:
@fisker I hacked the entries for `trailingComments` & `leadingComments` due to some type issues. I think these need to be fixed and will look into this. I did push these changes in case you may have any other ideas at the same time.

Reply:
have you try  `Array<PrettierEsComment | Comment>`?

Reply:
> `Array<PrettierEsComment | Comment>`

I think that would be no different from `Array<PrettierEsComment>`, since all members added to `PrettierEsComment` are optional.

I discovered that using `ReadonlyArray` is needed because of the types in `@babel/types`.

Reply:
I  checked our  comment attach logic, seems  we are  not  touching `leadingComments`, can  we  remove?

Reply:
yes (I did discover the same thing)